lib/std/fs.zig

```diff
@@ -369,6 +370,10 @@ pub const Dir = struct {
             fn nextSolaris(self: *Self) !?Entry {
                 start_over: while (true) {
                     if (self.index >= self.end_index) {
+                        if (self.first) {
+                            std.os.lseek_SET(self.dir.fs, 0) catch unreachable;
```

Comment:
Are we sure this cannot fail?

Reply:
It definitely can't on Linux, I assumed POSIX compliant systems would be the same, though I just pushed this to see how the tests run.

Reply:
I think an explanatory comment on each `catch unreacable` would be nice, but the possible errors for each platform are (from [`lseek(2)`](https://man7.org/linux/man-pages/man2/lseek.2.html)):

- `EBADF` when Dir.fd is invalid or wasn't opened with iteration. This `catch unreachable` would show up later anyway during the `next` call (see [here and other similar lines](https://github.com/ziglang/zig/blob/35423b005440c9f31a4cc6a53ccf6b7edd08b859/lib/std/fs.zig#L580)), so it being triggered during the `lseek_SET` just moves where it happens.
- `EINVAL` when "whence is not valid or the resulting file offset would be negative, or beyond the end of a seekable device.", but we're always using the known-to-be-valid `SEEK_SET` and the offset is always `0`
- `EOVERFLOW` when "The resulting file offset cannot be represented in an off_t", but we know `0` will be the resulting file offset
- `ENXIO` when lseek is using `SEEK_DATA` or `SEEK_HOLE` (actually, this isn't even a possible error within lseek_SET since we know we're using `SEEK_SET`)
- `ESPIPE` when "fd is associated with a pipe, socket, or FIFO.", but we know `fd` is a directory

WASI has an additional `ENOTCAPABLE` that I'm unsure about when/if it can be triggered.

Reply:
Hmm, I see. Curious to see if it applies to BSD/macOS too. I remember in the past I had some mixed experience with Dir iterators on macOS - the way the dir pointer works was completely different to how it works on Linux. But let's see if the tests pass. If so, then we're golden!

Reply:
Yeah, worth looking further into this.

Reply:
> I think an explanatory comment on each `catch unreachable` would be nice, but the possible errors for each platform are (from [`lseek(2)`](https://man7.org/linux/man-pages/man2/lseek.2.html)):
> 
> * `EBADF` when Dir.fd is invalid or wasn't opened with iteration. This `catch unreachable` would show up later anyway during the `next` call (see [here and other similar lines](https://github.com/ziglang/zig/blob/35423b005440c9f31a4cc6a53ccf6b7edd08b859/lib/std/fs.zig#L580)), so it being triggered during the `lseek_SET` just moves where it happens.
> * `EINVAL` when "whence is not valid or the resulting file offset would be negative, or beyond the end of a seekable device.", but we're always using the known-to-be-valid `SEEK_SET` and the offset is always `0`
> * `EOVERFLOW` when "The resulting file offset cannot be represented in an off_t", but we know `0` will be the resulting file offset
> * `ENXIO` when lseek is using `SEEK_DATA` or `SEEK_HOLE` (actually, this isn't even a possible error within lseek_SET since we know we're using `SEEK_SET`)
> * `ESPIPE` when "fd is associated with a pipe, socket, or FIFO.", but we know `fd` is a directory
> 
> WASI has an additional `ENOTCAPABLE` that I'm unsure about when/if it can be triggered.

`ENOTCAPABLE` would be triggered when we want iterate a dir tree we weren't given capabilities to do so.